compiler/src/org.graalvm.compiler.loop/src/org/graalvm/compiler/loop/InductionVariable.java

```diff
@@ -115,4 +115,15 @@ public ValueNode extremumNode() {
      * Deletes any nodes created within the scope of this object that have no usages.
      */
     public abstract void deleteUnusedNodes();
+
+    /**
+     * Range check predication support.
+     */
+    public abstract boolean isConstantScale();
```

Comment:
I'm looking for range check expressions of the form scale * i + offset in loop:
for (int i = ; i < ; i += stride) {
LoopPredicationPhase.verifyScaleAndOffset() implements logic that's quite similar with a chain of instanceof. It felt cleaner to implement it as methods of InductionVariable.
A sketch of that API is:
BasicInductionVariable.offset() is 0
BasicInductionVariable.scale() is 1
DerivedScaledInductionVariable.offset() is base.offset()
DerivedScaledInductionVariable.scale() is scale * base.scale()
DerivedOffsetInductionVariable.offset() is base.offset() + offset
DerivedOffsetInductionVariable.scale() is base.scale()

The API is a bit awkward because we would really need to return a 3 state value. For instance InductionVariable.offset() needs to return an offset node or a special value indicating a zero offset or a special value indicating that we can't tell the offset because the expression is too complicated. To achieve the same effect I split the 2 API points in 2:

offsetNode() returns either a node or null if it can't tell what the offset is
offsetIsZero() returns true if the offset is 0

Same with scale() with should return either a constant or some special value that tells scale is non constant or unknown.

Reply:
i think the issue comes from the fact that when you have a single induction variable "offset" and "scale" don't make much sense.

I think i see what you are trying to do here but without this context i can't really understand it that's why I think one option is to not put these method in the `InductionVariable` or to pass a second IV as argument so it's clear that it's an offset/scale relative to some other given IV.

Looking at the code in the phase i think the best is to pass the expected `BasicInductionVariable` to those functions so they can directly do the verification and there should be no need for `verifyScaleAndOffset`.

Reply:
Right that makes sense now. I updated the PR.